dali/pipeline/data/tensor_list.h

```diff
@@ -103,13 +112,28 @@ class DLL_PUBLIC TensorList : public Buffer<Backend> {
     }
     this->SetLayout(layout);
 
-    for (size_t i = 0; i < other.size(); ++i) {
-      type.template Copy<SrcBackend, Backend>(
-          raw_mutable_tensor(i),
-          other[i].raw_data(),
-          other[i].size(), stream);
-      this->meta_[i].SetSourceInfo(other[i].GetSourceInfo());
-      this->meta_[i].SetSkipSample(other[i].ShouldSkipSample());
+    if (use_copy_kernel && (std::is_same<SrcBackend, GPUBackend>::value || other.is_pinned())) {
+      if (!scatter_gather_) {
+        constexpr int kBlockSize = 1 << 19;  // 512kB per block
+        scatter_gather_.reset(new kernels::ScatterGatherGPU(kBlockSize, other.size()));
+      }
+      auto type_sz = type.size();
+      for (size_t i = 0; i < other.size(); ++i) {
+        scatter_gather_->AddCopy(raw_mutable_tensor(i), other[i].raw_data(),
+                                 other[i].size() * type.size());
+        this->meta_[i].SetSourceInfo(other[i].GetSourceInfo());
+        this->meta_[i].SetSkipSample(other[i].ShouldSkipSample());
+      }
+      scatter_gather_->Run(stream);
```

Comment:
I don't like the fact right now, that we actually have 2 places of implementation of actual Copy. One is here, inlined in the code and the second is in `types.cc`. But I don't have good idea what to do with this. Neither of them seems to be proper for it...